### PR TITLE
Samtools#312

### PR DESCRIPTION
**Pull request by jkbonfield**
The htslib part of the solution for issue samtools/samtools#312.

This adds an additional return value to sam_index_build / sam_index_build2.

### CODE REVIEW

File: cram/cram_index.c [hunk 1]
```diff
@@ -494,8 +494,8 @@ static int cram_index_build_multiref(cram_fd *fd,
  * fn_idx is the filename of the index file to be written;
  * if NULL, we add ".crai" to fn_base to get the index filename.
  *
- * Returns 0 on success
- *        -1 on failure
+ * Returns 0 on success,
+ *         negative on failure
```
Comment (mp15) [inline cram/cram_index.c]: It might be better to enumerate the list of different failures here.

Comment (jkbonfield) [inline cram/cram_index.c]: Done and force updated the commit.

File: sam.c [hunk 2]
```diff
@@ -522,6 +522,7 @@ int sam_index_build2(const char *fn, const char *fnidx, int min_shift)
         idx = bam_index(fp->fp.bgzf, min_shift);
         if (idx) {
             ret = hts_idx_save_as(idx, fn, fnidx, (min_shift > 0)? HTS_FMT_CSI : HTS_FMT_BAI);
+            ret = ret ? -4 : 0;
```
Comment (jmarshall) [inline sam.c]: Kinda incomprehensible.  Is the intent `if (ret < 0) ret = -4;` and is that clearer?

Comment (jkbonfield) [inline sam.c]: Essentially yes.  It's the analogue of if (ret != 0) ret = -4, but basically it's the same thing as the only working return value is zero.

I don't particularly find ?: incomprehensible, but if you feel strongly about it I can resubmit the patch with the change you suggest.

Comment (jmarshall) [inline sam.c]: Please do 😄

Me, I'm pro-`?:` in general, but in this case `ret = ret ? -4 : 0` bashes positive return values over the head, and to validate the change readers have to think about whether positive return values should be bashed to -4 (surely not!) or 0, and then go off and check exactly what `hts_idx_save_as` can return this week, and only then realise that positive return values (currently) won't occur.

Which is all a lot more work than just applying the general htslib rule that `retcode<0` ⇒ error.

Comment (jkbonfield) [inline sam.c]: Agreed.  At the very least it ought to ave been ret < 0 ? -4 : 0, but an "if" is easy enough too.

**Comment by jmarshall** [comment]
Thanks.  Squashed that last commit and merged as e1dd6353da4a93543674c447a12fccc95c167623.

**Pull request closed without merging**